File: python/tidegraph/token_window.py
"""Explicit LH token clock adapter; no executor calls or synthetic source rows."""
from .history import int64, increment
from .records import External


def token_inputs(outputs, continuation, layers, stop, *, body_cut, port=0):
    """Map complete body outputs [cut*layers, stop*layers) to readout token events.

    The caller seals the body interval and provides every output on the selected
    port. A globally empty token is outside original Pronounce's input domain.
    The continuation belongs to the readout graph; this function never mutates it.
    """
    q = continuation
    if (any(not int64(x) for x in (layers, stop, body_cut, port, q.cut, q.batch_size))
            or layers < 1 or q.batch_size < 1 or not 0 <= q.cut <= stop or port < 0
            or stop > (2**63-1)//layers or body_cut != stop*layers):
        raise ValueError("invalid or unsealed token window")
    selected, seen, tokens = [], set(), set()
    for batch, time, output_port, value in outputs:
        if not int64(output_port) or output_port < 0:
            raise ValueError("invalid body output port")
        if output_port != port:
            continue
        if (not int64(batch) or not int64(time) or not 0 <= batch < q.batch_size
                or not q.cut*layers <= time < body_cut or (batch, time) in seen):
            raise ValueError("invalid or duplicate body output coordinate")
        seen.add((batch, time)); token, phase = divmod(time, layers)
        selected.append((token, batch, phase, value)); tokens.add(token)
    if len(tokens) != stop-q.cut:
        raise ValueError("original Pronounce requires a nonempty global token window")
    positions = {}
    result = []
    for token, batch, phase, value in sorted(selected, key=lambda row: row[:3]):
        owner = batch, phase
        if owner not in positions:
            last = q.ledger.get(owner)
            if last is not None and (not isinstance(last, (tuple, list)) or len(last) != 2 or any(not int64(x) for x in last)
                                     or last[0] < 0 or not 0 <= last[1] < q.cut):
                raise ValueError("invalid token input ledger")
            positions[owner] = -1 if last is None else last[0]
        position = 0 if positions[owner] == -1 else increment(positions[owner])
        result.append(External(batch, phase, position, token, value))
        positions[owner] = position
    return result

File: cpp/src/token_window.cpp
#include "tide/token_window.h"
#include "tide/counters.h"
#include <algorithm>
#include <limits>
#include <set>
#include <stdexcept>

namespace tide {
std::vector<External> token_inputs(const std::vector<Output>& outputs, const Continuation& q,
                                  Index layers, Index stop, Index body_cut, Index port) {
  if (layers < 1 || q.batch_size < 1 || q.cut < 0 || stop < q.cut || port < 0
      || stop > std::numeric_limits<Index>::max()/layers || body_cut != stop*layers)
    throw std::invalid_argument("invalid or unsealed token window");
  std::vector<External> result;
  std::set<Owner> seen;
  std::set<Index> tokens;
  for (const auto& o : outputs) {
    if (o.port < 0) throw std::invalid_argument("invalid body output port");
    if (o.port != port) continue;
    if (o.batch < 0 || o.batch >= q.batch_size || o.time < q.cut*layers || o.time >= body_cut
        || !seen.emplace(o.batch, o.time).second)
      throw std::invalid_argument("invalid or duplicate body output coordinate");
    const auto token = o.time/layers, phase = o.time%layers;
    tokens.insert(token); result.push_back({o.batch, phase, 0, token, o.value});
  }
  if (tokens.size() != static_cast<size_t>(stop-q.cut))
    throw std::invalid_argument("original Pronounce requires a nonempty global token window");
  std::sort(result.begin(), result.end(), [](const auto& a, const auto& b) {
    return std::tie(a.time, a.batch, a.port) < std::tie(b.time, b.batch, b.port);
  });
  std::map<Owner, Index> positions;
  for (auto& x : result) {
    const Owner owner{x.batch, x.port};
    auto found = positions.find(owner);
    if (found == positions.end()) {
      const auto last = q.ledger.find(owner);
      if (last != q.ledger.end() && (last->second.first < 0 || last->second.second < 0 || last->second.second >= q.cut))
        throw std::invalid_argument("invalid token input ledger");
      found = positions.emplace(owner, last == q.ledger.end() ? -1 : last->second.first).first;
    }
    x.position = found->second == -1 ? 0 : increment(found->second);
    found->second = x.position;
  }
  return result;
}
}  // namespace tide
